main/src/ca/uwaterloo/flix/language/phase/Monomorph.scala

```diff
@@ -124,683 +124,683 @@ object Monomorph extends Phase[TypedAst.Root, TypedAst.Root] {
   // TODO: Monomorph: We use exceptions here as a temporary stop-gap. We should consider to restructure and use Validation.
 
   /**
-    * Performs monomorphization of the given AST `root`.
+    * A function-local queue of pending (fresh symbol, function definition, and substitution)-triples.
+    *
+    * For example, if the queue contains the entry:
+    *
+    * -   (f$1, f, [a -> Int])
+    *
+    * it means that the function definition f should be specialized w.r.t. the map [a -> Int] under the fresh name f$1.
     */
-  def run(root: Root)(implicit flix: Flix): Validation[Root, CompilationMessage] = flix.phase("Monomorph") {
-
-    // TODO: Monomorph: Un-nest these.
-
-    /**
-      * A function-local queue of pending (fresh symbol, function definition, and substitution)-triples.
-      *
-      * For example, if the queue contains the entry:
-      *
-      * -   (f$1, f, [a -> Int])
-      *
-      * it means that the function definition f should be specialized w.r.t. the map [a -> Int] under the fresh name f$1.
-      */
-    val defQueue: mutable.Queue[(Symbol.DefnSym, Def, StrictSubstitution)] = mutable.Queue.empty
+  private type DefQueue = mutable.Queue[(Symbol.DefnSym, Def, StrictSubstitution)]
 
-    /**
-      * A function-local map from a symbol and a concrete type to the fresh symbol for the specialized version of that function.
-      *
-      * For example, if the function:
-      *
-      * -   def fst[a, b](x: a, y: b): a = ...
-      *
-      * has been specialized w.r.t. to `Int` and `Str` then this map will contain an entry:
-      *
-      * -   (fst, (Int, Str) -> Int) -> fst$1
-      */
-    val def2def: mutable.Map[(Symbol.DefnSym, Type), Symbol.DefnSym] = mutable.Map.empty
-
-    /**
-      * Performs specialization of the given expression `exp0` under the environment `env0` w.r.t. the given substitution `subst0`.
-      *
-      * Replaces every reference to a parametric function with a reference to its specialized version.
-      *
-      * Replaces every local variable symbol with a fresh local variable symbol.
-      *
-      * If a specialized version of a function does not yet exists, a fresh symbol is created for it, and the
-      * definition and substitution is enqueued.
-      */
-    def specialize(exp0: Expression, env0: Map[Symbol.VarSym, Symbol.VarSym], subst0: StrictSubstitution): Expression = {
+  /**
+    * A function-local map from a symbol and a concrete type to the fresh symbol for the specialized version of that function.
+    *
+    * For example, if the function:
+    *
+    * -   def fst[a, b](x: a, y: b): a = ...
+    *
+    * has been specialized w.r.t. to `Int` and `Str` then this map will contain an entry:
+    *
+    * -   (fst, (Int, Str) -> Int) -> fst$1
+    */
+  private type Def2Def = mutable.Map[(Symbol.DefnSym, Type), Symbol.DefnSym]
```

Comment:
Sorry, the above comment applies here. Not to the type.

Reply:
What can we do? Pass it as an argument? Then we also need to pass the queue? So two arguments?

Reply:
Only the Def2Def/DefQueue types are declared in the object scope, each run makes its own two separate def2def and defQueue (157, 159) and passes them around.

Reply:
Ah yes, it was early and I am sleepy.